events/reporter.go

```diff
@@ -301,28 +301,37 @@ func SubscribeRewards() Subscription {
 
 // SubscribeToLayers is used to track and report automatically every time a
 // new layer is reached.
-func SubscribeToLayers(newLayerCh timesync.LayerTimer) {
+func SubscribeToLayers(ticker *timesync.Ticker) {
 	mu.RLock()
 	defer mu.RUnlock()
 
-	if reporter != nil {
-		// This will block, so run in a goroutine
-		go func() {
-			for {
-				mu.RLock()
-				stopChan := reporter.stopChan
-				mu.RUnlock()
-
-				select {
-				case layer := <-newLayerCh:
-					log.With().Debug("reporter got new layer", layer)
-					ReportNodeStatusUpdate()
-				case <-stopChan:
-					return
+	if reporter == nil {
+		return
+	}
+
+	// This will block, so run in a goroutine
+	go func() {
+		next := ticker.GetCurrentLayer().Add(1)
+		for {
+			mu.RLock()
+			stopChan := reporter.stopChan
+			mu.RUnlock()
+
+			select {
+			case <-ticker.AwaitLayer(next):
+				current := ticker.GetCurrentLayer()
+				if current.Before(next) {
+					log.Info("time sync detected, realigning ProposalBuilder")
```

Comment:
This applies here, too.
https://github.com/spacemeshos/go-spacemesh/pull/4013#discussion_r1089899203

Reply:
I had the following train of thought when I wrote that code:

1. `AwaitLayer` waits until at least `next`, but possibly longer (e.g. hibernation, etc.) so when it returns we have to check in which layer we are actually in.
2. `GetCurrentLayer` returns the current layer which is expected to be the one we waited for, but could be a later one, thats why we use `current` and not `layer` to avoid reporting an already passed layer/epoch.
3. In rare cases `current` is before `layer` (when exactly between `AwaitLayer` and `GetCurrentLayer` the time was adjusted backwards via NTP or by the user). In this case we must prevent reporting the same layer again.